docs/AQL/grammar/AqlParser.g4

```diff
@@ -0,0 +1,234 @@
+    ;
+
+
+selectExpr
+	: columnVar (AS IDENTIFIER)?
+	;
+
+fromExpr
+    : containsExpr
+    ;
+
+whereExpr
+    : NOT? identifiedExpr
+    | whereExpr AND whereExpr
+    | whereExpr OR whereExpr
+    | OPEN_PAR whereExpr CLOSE_PAR
+    ;
+
+orderByExpr
+	: identifiedPath order=(DESCENDING|DESC|ASCENDING|ASC)?
+	;
+
+
+columnVar
+	: identifiedPath
+	| aggregateFunctionCall
+	| functionCall
+	;
+
+containsExpr
+    : classExprOperand (CONTAINS containsExpr)?
+    | containsExpr AND containsExpr
+    | containsExpr OR containsExpr
+    | OPEN_PAR containsExpr CLOSE_PAR
+    ;
+
+identifiedExpr
+    : EXISTS identifiedPath
+    | identifiedPath COMPARISON_OPERATOR identifiedOperand
+    | functionCall COMPARISON_OPERATOR identifiedOperand
+    | identifiedPath LIKE likeOperand
+    | identifiedPath MATCHES OPEN_CURLY matchesOperand CLOSE_CURLY
+    ;
+
+identifiedOperand
+ 	: primitive
+ 	| PARAMETER
+ 	| identifiedPath
+ 	| functionCall
+ 	;
+
+identifiedPath
+    : IDENTIFIER pathPredicate? (SLASH objectPath)?
+    ;
+
+pathPredicate
+ 	: OPEN_BRACKET (standardPredicate | archetypePredicate | nodePredicate) CLOSE_BRACKET
+ 	;
+
+standardPredicate
+    : objectPath COMPARISON_OPERATOR pathPredicateOperand
+    ;
+
+archetypePredicate
+    : ARCHETYPEID
+    | PARAMETER
+    ;
+
+nodePredicate
+    : NODEID (COMMA (STRING|PARAMETER))?
+    | ARCHETYPEID (COMMA (STRING|PARAMETER))?
+    | PARAMETER
+    | objectPath COMPARISON_OPERATOR pathPredicateOperand
+    | objectPath MATCHES REGEXPATTERN
+    | nodePredicate AND nodePredicate
+    | nodePredicate OR nodePredicate
+    ;
+
+pathPredicateOperand
+    : primitive
+ 	| objectPath
+ 	| PARAMETER
+ 	;
+
+
+objectPath
+ 	: pathPart (SLASH pathPart)*
+ 	;
+pathPart
+ 	: IDENTIFIER pathPredicate?
+ 	;
+
+likeOperand
+    : STRING
+    | PARAMETER
+    ;
+matchesOperand
+ 	: valueListItem (COMMA valueListItem)*
+ 	| terminologyFunction
+ 	| URIVALUE
+ 	;
+valueListItem
+ 	: primitive
+ 	| PARAMETER
+ 	| terminologyFunction
+ 	;
+
+primitive
+    : STRING
+    | INTEGER
+    | FLOAT
+    | DOUBLE
+    | DATE
+    | BOOLEAN
+    ;
+
+
+
+classExprOperand
+	: IDENTIFIER IDENTIFIER? // RM_TYPE_NAME variable
+    | archetypedClassExpr
+    | versionedClassExpr
+	| versionClassExpr
+	;
+
+// RM_TYPE_NAME [archetype_id]
+// RM_TYPE_NAME variable [archetype_id]
+archetypedClassExpr
+ 	: IDENTIFIER IDENTIFIER? archetypePredicate
+ 	;
+versionedClassExpr
+ 	: VERSIONED_OBJECT IDENTIFIER? (OPEN_BRACKET standardPredicate OPEN_BRACKET)?
+ 	;
+versionClassExpr
+ 	: VERSION IDENTIFIER? (OPEN_BRACKET (standardPredicate|versionPredicate) OPEN_BRACKET)?
+ 	;
+versionPredicate
+ 	: LATEST_VERSION
+ 	| ALL_VERSIONS
+ 	;
+
+functionCall
+    : terminologyFunction
+    | function
+    ;
+
+function
+    : IDENTIFIER OPEN_PAR functionArg (COMMA functionArg)* CLOSE_PAR
+    ;
+
+functionArg
+    : primitive
+    | identifiedPath
+    | PARAMETER
+    | functionCall
+    ;
+
+aggregateFunctionCall
```

Comment:
I don't see a real need here to add separate rules for each function definition. I'd suggest simply doing:
```
aggregateFunctionCall 
  : COUNT OPEN_PAR (DISTINCT? identifiedPath | STAR) CLOSE_PAR
  | MIN OPEN_PAR identifiedPath CLOSE_PAR
  | MAX OPEN_PAR identifiedPath CLOSE_PAR
  | SUM OPEN_PAR identifiedPath CLOSE_PAR
  | AVG OPEN_PAR identifiedPath CLOSE_PAR
  ;
```
The extra indirection only deepens the parse tree, and while this is generally not an issue, it doesn't really add much here, except maybe human readability. If you want to make things slightly more clear, you could add alternative names like this:
```
aggregateFunctionCall 
  : COUNT OPEN_PAR (DISTINCT? identifiedPath | STAR) CLOSE_PAR #countFunction
  | MIN OPEN_PAR identifiedPath CLOSE_PAR                      #minFunction
  | MAX OPEN_PAR identifiedPath CLOSE_PAR                      #maxFunction
  | SUM OPEN_PAR identifiedPath CLOSE_PAR                      #sumFunction
  | AVG OPEN_PAR identifiedPath CLOSE_PAR                      #avgFunction
  ;
```
Although implementers may or may not want to use those.

Reply:
good suggestion, I'll do it